core/services/pipeline/common.go

```diff
@@ -168,17 +170,30 @@ type JSONSerializable struct {
 	Null bool
 }
 
+// UnmarshalJSON implements custom unmarshaling logic
 func (js *JSONSerializable) UnmarshalJSON(bs []byte) error {
 	if js == nil {
 		*js = JSONSerializable{}
 	}
+
 	return json.Unmarshal(bs, &js.Val)
 }
 
+// MarshalJSON implements custom marshaling logic
 func (js JSONSerializable) MarshalJSON() ([]byte, error) {
 	switch x := js.Val.(type) {
 	case []byte:
-		return json.Marshal(string(x))
+		// Don't need to HEX encode if it is a valid JSON string
+		if json.Valid(x) {
+			return json.Marshal(string(x))
+		}
+
+		// Don't need to HEX encode if it is already HEX encoded value
+		if _, err := hex.DecodeString(strings.TrimPrefix(string(x), "0x")); err == nil {
+			return json.Marshal(string(x))
+		}
+
+		return json.Marshal(hex.EncodeToString(x))
```

Comment:
This will need a similar workaround in `UnmarshalJSON` in order to properly round-trip. 

\cc @spooktheducks WDYT about this approach?

Reply:
@archseer we need to know which value should be decoded and which should not. That is why I've added a prefix in my initial implementation, but @samsondav doesn't like that.